### PR TITLE
src: Sharding cleanup and checkReady rewrite

### PR DESCRIPTION
**Pull request by vladfrangu**
**Please describe the changes this PR makes and why it should be merged:**

This PR does a bit of cleanup on the internal sharding code, as well as rewrite the *poorly done* checkReady method. Also updates a few modules (gotta stay up to date 💪) as well as reduces a bit of code dupe

### But Vlad...why?! Why change what isn't broken?

Good question. While most of the changes here are not breaking (like the code de-dupe from WebSocketShard#connect), others may be (like the upcoming checkReady rewrite). While I do test everything with some *you wouldn't imagine anyone would do this but just in case* methods, I still cannot cover ALL edge cases, so, as always, if you wanna test it, `npm i vladfrangu/discOwOd.js#sharding-cleanup` (or `yarn add`, you know the drill by now) and report back with issues! ❤️ 

### Ok, ok.. Tell us what's changed

Gee, no need to rush in...

- Bump eslint's ecmaVersion to 2019 from 2018 (so try-catch without the error aren't considered errors by eslint)
- Document utf-8-validate as a module that you may use to improve the ws performance
- ~~Bump pako to 1.0.10~~ Remove pako as a dependency
  - Use zlib-sync on node envs if you want compression
- Bump ws to 7.2.0 (no breaking changes that affect us, phew)
- Convert var usage to let in WebSocket.js
	- Only one `var` remains... Soon to be gone 👀 
- Merge `totalShardCount` in `shardCount`
	- You now set shards to a number, an array of numbers or 'auto'
- Reduced some code dupe in WebSocketShard#connect
- ~~Minor bugfixing for uws possibly~~ Completely removed uws! Buh-bye module that made me rip my hair out!
- Rewrote checkReady, it is now per-shard, and allows the client to "ready" even if some guilds are unavailable

### TODO

- ~~Move the `auto` property to `shards`, not `totalShardCount`~~
- ~~Rewrite, test, break checkReady~~
- ~~Complain to our local snake about things~~
- ~~Mark this PR as ready to review once it's ready for review~~

**Status**
- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [x] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by iCrawl** [comment]
Might as well remove every single trace of uWs, not worth to bring an old version up to date anymore.

**Comment by vladfrangu** [comment]
As of commit https://github.com/discordjs/discord.js/pull/3393/commits/b2c15fc205231b5b015db9fd3dd3c1a0ff675e1f and https://github.com/discordjs/discord.js/pull/3393/commits/dba28d64c6d7e8c54a524a962895e8dba8ca625c,

- The last part of the bot's token is censored. Considering we request the debug logs for certain WS issues, some people might accidentally copy their token in the logs. We censor the last part (which is the actual cryptographic part, useless on its own)
- Setting shards to `auto` spawns as many shards as Discord suggests

## So, what happens with different sharding configurations?

### Setting shards to `auto`

```js
new Client({ shards: 'auto' });
```

Spawns as many shards as Discord suggests

### Setting totalShardCount to a number

```js
new Client({ totalShardCount: 4 });
```

Spawns 4 shards

### Setting shards to a number, and totalShardCount to a number

```js
new Client({ shards: 0, totalShardCount: 2 });
```

Spawns shard 0/2 (normalized 1/2)

### Setting shards to an array of numbers, and totalShardCount to a number

```js
new Client({ shards: [0, 1], totalShardCount: 4 });
```

Spawns shard 0 and 1 out of 4 (normalized 1/4 and 2/4)

### Setting shards to `auto` and totalShardCount to a number

```js
new Client({ shards: 'auto', totalShardCount: 12 });
```

Spawns as many shards suggested by Discord. This will IGNORE totalShardCount.

#### Discussion

Do we want to change that? In my opinion if you set it to `auto`, you expect it to handle everything already, 

### Setting shards to anything other than `auto`, a number or an array of valid shard IDs (no negative numbers, Infinity, floats, NaN's, nulls, strings, objects, snakes, or blood of your enemies)

```js
new Client({
	shards: [null, {}, 0.25, "enemy's blood", -Infinity, NaN],
	totalShardCount: 4,
});
```

Yields

```
PS C:\Users\Vlad\Desktop\discOwOd.js\test> node vladtester9001
C:\Users\Vlad\Desktop\discOwOd.js\src\client\Client.js:376
    if (options.shards && !options.shards.length) throw new RangeError('CLIENT_INVALID_PROVIDED_SHARDS');
                                                  ^

RangeError [CLIENT_INVALID_PROVIDED_SHARDS]: None of the provided shards were valid.
    at Client._validateOptions (C:\Users\Vlad\Desktop\discOwOd.js\src\client\Client.js:376:57)
    at new Client (C:\Users\Vlad\Desktop\discOwOd.js\src\client\Client.js:71:10)
    at Object.<anonymous> (C:\Users\Vlad\Desktop\discOwOd.js\test\vladtester9001.js:5:16)
    at Module._compile (internal/modules/cjs/loader.js:774:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:785:10)
    at Module.load (internal/modules/cjs/loader.js:641:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:837:10)
    at internal/main/run_main_module.js:17:11 {
  [Symbol(code)]: 'CLIENT_INVALID_PROVIDED_SHARDS'
}
```

So.. don't.

**Comment by appellation** [comment]
I would think that we should exit if *any* of the shards are not valid, since that indicates some sort of undefined state.

It might be worth considering a sort of "shard count manager" to abstract some of this logic out of the client and websocket manager. This could simplify the client options and help keep code more explicit. A quick hypothetical: `new Client({ shards: new ShardCountManager([0, 3, 6], 9) })`.

An alternative might be exposing the websocket manager or whatever is responsible for spawning the shards, reworking the constructor parameters, and allowing people to pass custom ones through the client option. This would look pretty similar to my original concept but the class given in the options would be responsible for more than just shard counting. This would expose a very powerful API for custom sharding solutions but would obviously also present more opportunities for users to introduce significant issues with their sharding.

**Comment by xonrsoftware** [comment]
Using this with all additional packages deleted. Finally d.js working without silent disconnect. Thanks.

**Comment by vladfrangu** [comment]
@xonrsoftware glad to hear that!

Can you tell me what issues you've encountered with the optional packages? I'd love to know them, in case there are other bugs to be fixed!

**Comment by xonrsoftware** [comment]
> @xonrsoftware glad to hear that!
> 
> Can you tell me what issues you've encountered with the optional packages? I'd love to know them, in case there are other bugs to be fixed!

I just deleted them before started to using this branch. I specify it for others, who may still experience silent disconnects using this branch.

File: src/client/Client.js [hunk 1]
```diff
@@ -363,14 +365,15 @@ class Client extends BaseClient {
    * @private
    */
   _validateOptions(options = this.options) { // eslint-disable-line complexity
-    if (options.shardCount !== 'auto' && (typeof options.shardCount !== 'number' || isNaN(options.shardCount))) {
-      throw new TypeError('CLIENT_INVALID_OPTION', 'shardCount', 'a number or "auto"');
+    if (typeof options.shardCount !== 'number' || isNaN(options.shardCount) || options.shardCount < 1) {
+      throw new TypeError('CLIENT_INVALID_OPTION', 'shardCount', 'a number greater than 1 or "auto"');
```
Comment (bdistin) [inline src/client/Client.js]:
```suggestion
      throw new TypeError('CLIENT_INVALID_OPTION', 'shardCount', 'a number greater than 1');
```

File: src/client/websocket/WebSocketManager.js [hunk 2]
```diff
@@ -396,45 +396,37 @@ class WebSocketManager extends EventEmitter {
...
-          this.debug(`Failed to fetch all members before ready! ${e}\n${e.stack}`);
-          this.triggerReady();
+    this.status = Status.NEARLY;
+
+    if (this.client.options.fetchAllMembers) {
+      try {
+        const promises = this.client.guilds.map(guild => {
+          if (guild.available) return guild.members.fetch();
+          // Return no-op if guild is unavailable
+          return () => null;
```
[the remark this answers is not in the document]
Comment (vladfrangu) [reply src/client/websocket/WebSocketManager.js]: Technically, both `Promise.resolve()` and the current code are valid, as it's there just to ensure `Promise.all` gets functions. I can do the change soon!

File: src/client/websocket/WebSocketShard.js [hunk 3]
```diff
@@ -128,24 +125,28 @@ class WebSocketShard extends EventEmitter {
...
-   * Average heartbeat ping of the websocket, obtained by averaging the WebSocketShard#pings property
-   * @type {number}
-   * @readonly
-   */
-  get ping() {
-    const sum = this.pings.reduce((a, b) => a + b, 0);
-    return sum / this.pings.length;
+    /**
+     * A set of guild IDs this shard expects to receive
+     * @type {Set<string>}
```
Comment (kyranet) [inline src/client/websocket/WebSocketShard.js]: This property is initially set as `undefined`, per line 142
```suggestion
     * @type {?Set<string>}
```

File: src/client/websocket/WebSocketShard.js [hunk 4]
```diff
@@ -324,7 +310,7 @@ class WebSocketShard extends EventEmitter {
    * @private
    */
   onClose(event) {
-    this.closeSequence = this.sequence;
+    if (this.sequence && this.sequence !== -1) this.closeSequence = this.sequence;
```
Comment (kyranet) [inline src/client/websocket/WebSocketShard.js]: `this.sequence` can only be falsy when it is `0` (considering all assignments are to numbers), and it is redudant with `this.sequence !== -1`, since it cannot be falsy and `-1`:
```suggestion
    if (this.sequence !== -1) this.closeSequence = this.sequence;
```

File: typings/index.d.ts [hunk 5]
```diff
@@ -1666,12 +1667,13 @@ declare module 'discord.js' {
 		private connection: WebSocket | null;
 		private helloTimeout: NodeJS.Timeout | null;
 		private eventsAttached: boolean;
+		private expectedGuilds: Set<Snowflake>;
```
Comment (kyranet) [inline typings/index.d.ts]: This property is either a Set<T> or undefined: https://github.com/discordjs/discord.js/blob/e47fc1e0f85ee63409979084bd45e0b748899fad/src/client/websocket/WebSocketShard.js#L128

```suggestion
		private expectedGuilds: Set<Snowflake> | undefined;
```

File: typings/index.d.ts [hunk 6]
```diff
@@ -1666,12 +1667,13 @@ declare module 'discord.js' {
...
+		private readyTimeout: NodeJS.Timeout | null;
```
Comment (kyranet) [inline typings/index.d.ts]: This property is either a NodeJS.Timeout or undefined: https://github.com/discordjs/discord.js/blob/e47fc1e0f85ee63409979084bd45e0b748899fad/src/client/websocket/WebSocketShard.js#L149

```suggestion
		private readyTimeout: NodeJS.Timeout | undefined;
```

File: src/client/websocket/WebSocketShard.js [hunk 7]
```diff
@@ -417,11 +406,49 @@ class WebSocketShard extends EventEmitter {
...
     }
+    // Step 2. Create a 15s timeout that will mark the shard as ready if there are still unavailable guilds
+    this.readyTimeout = this.manager.client.setTimeout(() => {
+      this.debug(`Shard did not receive any more guild packets in 15 seconds.
+  Unavailable guild count: ${this.expectedGuilds.size}`);
+
+      this.status = Status.READY;
+
+      this.emit(ShardEvents.ALL_READY, this.expectedGuilds);
+    }, 15000);
```
Comment (kyranet) [inline src/client/websocket/WebSocketShard.js]: `this.readyTimeout` is only used thrice in this file, here in line 444 and in https://github.com/discordjs/discord.js/blob/e47fc1e0f85ee63409979084bd45e0b748899fad/src/client/websocket/WebSocketShard.js#L426

Which checks if it's truthy (since it's either undefined or `NodeJS.Timeout`), but once this is set, it's never set to `undefined`, thus running `clearTimeout` every time `checkReady` is called and this has been set.

My guess is that you can assign `this.readyTimeout = undefined;` inside the callback so it is freed automatically, here as well as in line 426.

File: typings/index.d.ts [hunk 8]
```diff
@@ -1666,12 +1667,13 @@ declare module 'discord.js' {
 		private connection: WebSocket | null;
 		private helloTimeout: NodeJS.Timeout | null;
```
Comment (kyranet) [inline typings/index.d.ts]: This line needs to be edited due to the change in https://github.com/discordjs/discord.js/blob/e47fc1e0f85ee63409979084bd45e0b748899fad/src/client/websocket/WebSocketShard.js#L128
```suggestion
		private helloTimeout: NodeJS.Timeout | undefined;
```

**Comment by gc** [comment]
Been running this for a few weeks with no issues 👍

**Comment by pq987** [comment]
whenever i startup the bot it never actually starts up and is instead is stuck in an infinite loop of trying to login here are the debug logs and error that is infinitely logged/thrown
```
Provided token: ***************************
Preparing to connect to the gateway...
[WS => Manager] Manager was destroyed. Called by:
Error: MANAGER_DESTROYED
    at WebSocketManager.destroy (/root/Vortex/node_modules/discord.js/src/client/websocket/WebSocketManager.js:340:54)
    at Vortex.destroy (/root/Vortex/node_modules/discord.js/src/client/Client.js:226:13)
    at Vortex.login (/root/Vortex/node_modules/discord.js/src/client/Client.js:215:12)
    at process._tickCallback (internal/process/next_tick.js:68:7)
{ AbortError: The user aborted a request.
    at RequestHandler.execute (/root/Vortex/node_modules/discord.js/src/rest/RequestHandler.js:106:9)
    at process._tickCallback (internal/process/next_tick.js:68:7)
  name: 'AbortError',
  code: 500,
  method: 'get',
  path: '/gateway/bot' }
```
to try get around the user abort error i manually removed the lines where it would abort the request after a timeout but by doing this the cpu spikes to 100%... causing the server to crash.
